### PR TITLE
WIP: Initial work on adding packml_stacklight

### PR DESCRIPTION
**Pull request by winchesterag**
- [x] code / design review
- [x] add config/startup parameters and node testing
- [x] test subscribe and publish
- [x] test actual node with hardware

### CODE REVIEW

File: packml_stacklight/include/packml_stacklight/utils.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+enum LightValue
+{
+  UNDEFINED = 0,
+  RED = 1,
+  AMBER = 2,
+  GREEN = 3,
+  BLUE = 4,
+};
+
+static std::map<LightValues::LightValue, std::string> LightValueMap = { { UNDEFINED, "UNDEFINED-LIGHT" },
```
Comment (shaun-edwards) [inline packml_stacklight/include/packml_stacklight/utils.h]: Variable naming doesn't meet style guide.

File: packml_stacklight/src/packml_stacklight.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ */
+
+#include "packml_stacklight/packml_stacklight.h"
+#include "packml_stacklight/utils.h"
+
+namespace packml_stacklight
+{
+PackmlStacklight::PackmlStacklight(ros::NodeHandle nh, ros::NodeHandle pn) : nh_(nh), pn_(pn)
+{
+  ros::NodeHandle packml_node("~/packml");
```
Comment (shaun-edwards) [inline packml_stacklight/src/packml_stacklight.cpp]: This is a bit odd....you don't normally need a reference to another node.

File: packml_stacklight/src/packml_stacklight.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  status_sub_ = packml_node.subscribe<packml_msgs::Status>("status", 1, &PackmlStacklight::callBackStatus, this,
```
Comment (shaun-edwards) [inline packml_stacklight/src/packml_stacklight.cpp]: Any subscribers should use the passed in `nh`

Comment (shaun-edwards) [reply packml_stacklight/src/packml_stacklight.cpp]: `ros::TransportHints().reliable().tcpNoDelay(true)` should only be used on high frequency topics.

File: packml_stacklight/src/packml_stacklight.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                                                           ros::TransportHints().reliable().tcpNoDelay(true));
+
+  packml_msgs::State temp;
+  temp.val = packml_msgs::State::UNDEFINED;
+  StatusAction* status_ptr = getActionFromState(temp);
```
Comment (shaun-edwards) [inline packml_stacklight/src/packml_stacklight.cpp]: Does this need to be a pointer?  We should only use pointers where required?

File: packml_stacklight/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+## Add gtest based cpp test target and link libraries
+#if(TARGET ${PROJECT_NAME}_test)
+if(CATKIN_ENABLE_TESTING)
+	find_package(rostest)
+	catkin_add_gtest(${PROJECT_NAME}_test test/launch/test_stacklight.launch test/test_stacklight.cpp)
+	target_link_libraries(${PROJECT_NAME}_test ${PROJECT_NAME} ${catkin_LIBRARIES})
+endif()
+
+#if(CATKIN_ENABLE_TESTING)
```
Comment (shaun-edwards) [inline packml_stacklight/CMakeLists.txt]: Please delete commented items

File: packml_stacklight/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+#  find_package(rostest)
+#  set(UTEST_SRC_FILES test/test_stacklight.cpp)
+#
+#  catkin_add_gtest(${PROJECT_NAME}_utest ${UTEST_SRC_FILES})
+#  target_compile_options(${PROJECT_NAME}_utest PUBLIC -std=c++11)
+#  target_link_libraries(${PROJECT_NAME}_utest ${PROJECT_NAME} ${catkin_LIBRARIES})
+#endif()
+
+#install(TARGETS ${PROJECT_NAME}
```
Comment (shaun-edwards) [inline packml_stacklight/CMakeLists.txt]: Install commands are required

**Review by shaun-edwards — CHANGES_REQUESTED** [review]
@winchesterag, just a couple of items to clean up.  Overall, looks good!

File: packml_stacklight/CMakeLists.txt (hunk 6 above)
Comment (winchesterag) [reply packml_stacklight/CMakeLists.txt]: I have added in the install commands in, but not sure how to test.
I also am not familiar with what is needed, however used what other packages had.

**Comment by 130s** [comment]
CI failing. E.g. https://travis-ci.org/ros-industrial-consortium/packml/jobs/539897337#L2317

```
/root/catkin_ws/devel/.private/packml_stacklight/lib/packml_stacklight/packml_stacklight_utest --gtest_output=xml:/root/catkin_ws/build/packml_stacklight/test_results/packml_stacklight/gtest-packml_stacklight_utest.xml
[FATAL] [1559338761.464739287]: ROS_MASTER_URI is not defined in the environment. Either type the following or (preferrably) add this to your ~/.bashrc file in order set up your local machine as a ROS master:
export ROS_MASTER_URI=http://localhost:11311
then, type 'roscore' in another shell to actually launch the master program.
Segmentation fault (core dumped)
:
build/packml_stacklight/test_results/packml_stacklight/MISSING-gtest-packml_stacklight_utest.xml: 1 tests, 0 errors, 1 failures, 0 skipped
Summary: 6 tests, 0 errors, 1 failures, 0 skipped
/root/ici/tests/source_tests.sh: line 244: error: command not found
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
Function 'catkin_run_tests' returned with code '127' after 0 min 19 sec
```
Direct cause is I think a test case code looks for ROS Master but `ROS_MASTER_URI` environment variable, which tells your ROS programs the location of the Master, is not found. That happened probably because there's `ros::init` at [packml_stacklight/test/test_stacklight.cpp](https://github.com/ros-industrial-consortium/packml/blob/8cf05d33f7ce4eb2b47599883fb0b6d87ae2d099/packml_stacklight/test/test_stacklight.cpp#L768), which I think tries to connect to ROS Master. There seems to be a launch file for testing ([test_stacklight.launch](https://github.com/ros-industrial-consortium/packml/blob/ef489bea7eaeab9c1dedd3bb73852e469a6d14f5/packml_stacklight/test/launch/test_stacklight.launch)), which starts ROS Master, but in your [testing setting the launch file isn't used](https://github.com/ros-industrial-consortium/packml/blob/ef489bea7eaeab9c1dedd3bb73852e469a6d14f5/packml_stacklight/CMakeLists.txt#L63). If you run the launch file, at least you'll see something differently as ROS Master should start.

Configuring `rostest` and reading its log is tricky, but [this on docs.ros.org](http://docs.ros.org/kinetic/api/catkin/html/howto/format2/rostest_configuration.html) is the resource I always go back to.

**Pull request merged**